### PR TITLE
fix(cdis_refresh_token): Fix mising of refresh_token for cdis backend

### PR DESCRIPTION
**Pull request by giangbui**
Added new table 'refresh_token' to DB
- insert new refresh token whenever new token is generated
- query the table to obtain refresh token

@thanh-nguyen-dang , @rudyardrichter , @philloooo

### CODE REVIEW

**Comment by giangbui** [comment]
resolve #24

File: tests/conftest.py [hunk 1]
```diff
@@ -138,12 +138,12 @@ def app(request):
     root_dir = os.path.dirname(os.path.realpath(__file__))
     app_init(fence.app, test_settings, root_dir=root_dir)
 
-    def fin():
-        for tbl in reversed(Base.metadata.sorted_tables):
-            fence.app.db.engine.execute(tbl.delete())
-        mocker.unmock_functions()
+    # def fin():
```
Comment (rudyardrichter) [inline tests/conftest.py]: Why is this removed?

File: fence/blueprints/storage_creds.py [hunk 2]
```diff
@@ -66,6 +68,10 @@ def list_credentials(backend):
     if backend != 'cdis':
         result = capp.storage_manager.list_keypairs(backend, g.user)
         return jsonify(dict(access_keys=result))
+    if backend == 'cdis':
+        with capp.db.session as session:
+            tokens = session.query(RefreshToken.token).filter_by(userid=g.user.id).all()
+            return jsonify(dict(token=tokens))
```
Comment (rudyardrichter) [inline fence/blueprints/storage_creds.py]: this would look cleaner with the conditionals reordered as:
```python
if backend == 'cdis':
    with capp.db.session as session:
        # ...
else:
    result = capp.storage_manager.list_keypairs(backend, g.user)
    # ...
```

File: fence/data_model/models.py [hunk 3]
```diff
@@ -68,6 +68,18 @@ def validate_scopes(self, scopes):
...
+    __tablename__ = "refresh_token"
+
+    token = Column(String, primary_key=True)
+    userid = Column(Integer)
+
+    def delete(self):
+        with capp.db.session as session:
+            session.delete(self)
+            session.commit()
+            return self
```
Comment (rudyardrichter) [inline fence/data_model/models.py]: What is the `return self` for?

File: fence/blueprints/storage_creds.py (hunk 2 above)
Comment (giangbui) [reply fence/blueprints/storage_creds.py]: fixed!

File: fence/data_model/models.py [hunk 4]
```diff
@@ -68,6 +70,18 @@ def validate_scopes(self, scopes):
         scopes = scopes[0].split(',')
         return all(scope in self._scopes for scope in scopes)
 
+class RefreshToken(Base):
+    __tablename__ = "refresh_token"
+
+    token = Column(String, primary_key=True)
+    userid = Column(Integer)
+    created_time = Column(DateTime, default=datetime.datetime.utcnow().isoformat)
```
Comment (rudyardrichter) [inline fence/data_model/models.py]: Rather than a `created_time` it would be better to get the `exp` (expiration) field from the token so that cleaning out old entries from the table can just check that column against the current unix time.

File: fence/data_model/models.py [hunk 5]
```diff
@@ -68,6 +70,18 @@ def validate_scopes(self, scopes):
         scopes = scopes[0].split(',')
         return all(scope in self._scopes for scope in scopes)
 
+class RefreshToken(Base):
+    __tablename__ = "refresh_token"
+
+    token = Column(String, primary_key=True)
```
Comment (rudyardrichter) [inline fence/data_model/models.py]: Similarly to the `exp` thing, the tokens have a `jti` (JWT ID) field which would make for a better primary key than just using the entire token.

**Comment by giangbui** [comment]
Codacy comments the UserRefreshToken is unused but it is used in storage_creds.py. Probably, it is codacy bug.

File: fence/blueprints/storage_creds.py [hunk 6]
```diff
@@ -69,7 +71,12 @@ def list_keypairs(provider):
         }
 
     '''
-    if provider != 'cdis':
+    if provider == 'cdis':
```
Comment (thanh-nguyen-dang) [inline fence/blueprints/storage_creds.py]: should add an example of return results so that people use it can know about the format.
